File: axe-playwright-example/utils/HightLighter.ts
import {Page} from 'playwright';
import {Locator, test} from "@playwright/test";
import {AxeScanResult, Impact} from "./types";

class Highlighter {
    private readonly page: Page;

    constructor(page: Page) {
        this.page = page;
    }

    async highlightElements(data: AxeScanResult): Promise<void> {
        let globalCounter = 0;
        for (const item of data.violations) {
            for(const node of item.nodes)  {
                await this.highlightElement(++globalCounter, this.page.locator(node.target[0].toString()), this.page, node.impact);
            }
        }
        const screenshot = await this.page.screenshot({fullPage: true});
        await test.info().attach('Violations on page', {body: screenshot, contentType: 'image/png'});
    }

    async removeHighlights(): Promise<void> {
        await this.page.evaluate(() => {
            document.querySelectorAll('.violations-border').forEach(
                el => el.remove()
            );
            document.querySelectorAll('.violations-index').forEach(
                el => el.remove()
            );
        });
    }

    private async highlightElement(index: number, element: Locator, page: Page, impact: Impact) {
        const boundingBox = await element.boundingBox();
        const color = this.getImpactColor(impact);

        await page.evaluate(({
                                 boundingBox,
                                 index,
                                 color}) => {
            const customDiv = document.createElement('div');
            customDiv.className = 'violations-border';
            customDiv.style.position = 'absolute';
            customDiv.style.left = `${boundingBox?.x ?? 0}px`;
            customDiv.style.top = `${boundingBox?.y ?? 0}px`;
            customDiv.style.width = `${boundingBox?.width ?? 0}px`;
            customDiv.style.height = `${boundingBox?.height ?? 0}px`;
            customDiv.style.border = `2px solid ${color}`;

            document.body.appendChild(customDiv);

            const customSpan = document.createElement('b');
            customSpan.className = 'violations-index';
            customSpan.style.position = 'absolute';
            customSpan.innerText = index + '';
            customSpan.style.color = color;
            customSpan.style.left = `${boundingBox?.x - 20}px`;
            customSpan.style.top = `${boundingBox?.y}px`;

            document.body.appendChild(customDiv);
            document.body.appendChild(customSpan);

        }, {boundingBox, index, color});
    }

    private getImpactColor = (impact: Impact) => {
        switch(impact){
            case "critical":
                return "#da1e28";
            case "serious":
                return "#ff832b";
            case "moderate":
                return "#f1c21b";
            case "minor":
                return "#add8e6";
            default: throw new Error("Unknown impact");
        }
    }
}

export default Highlighter;

File: axe-selenide/src/test/java/Hightlighter.java
import com.codeborne.selenide.SelenideElement;
import com.codeborne.selenide.WebDriverRunner;
import com.deque.html.axecore.results.CheckedNode;
import com.deque.html.axecore.results.Results;
import com.deque.html.axecore.results.Rule;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import io.qameta.allure.Allure;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import java.io.ByteArrayInputStream;
import java.util.*;

import static com.codeborne.selenide.Selenide.$;
import static com.codeborne.selenide.Selenide.executeJavaScript;

public class Hightlighter {
    private List<SelenideElement> els = new ArrayList<>();
    private int globalCounter = 0;

    public void highlightElements(Results data) {
        for (Rule item : data.getViolations()) {
            for (CheckedNode node : item.getNodes()) {
                String selector = node.getTarget().toString();
                String substring = selector.substring(1, selector.length() - 1);
                els.add($(substring));
            }
        }
        for (SelenideElement el : els) {
            highlightElement(++globalCounter, el, Impact.CRITICAL);
        }
        Allure.addAttachment(
                "problems on screen",
                new ByteArrayInputStream(
                        takeScreenshot(WebDriverRunner.getWebDriver()))
        );
        removeHighlights();
    }

    public void removeHighlights() {
        for (SelenideElement el : els) {
            executeJavaScript("arguments[0].style.border='none'", el);
        }
        executeJavaScript("document.querySelectorAll('.violations-index').forEach(el => el.remove());");
    }

    private void highlightElement(int index, SelenideElement element, Impact impact) {
        executeJavaScript("arguments[0].style.border='2px solid red'", element);
        executeJavaScript("arguments[0].setAttribute('data-index', arguments[1])", element, index);
        executeJavaScript("arguments[0].innerHTML = arguments[0].innerHTML + '<span class=\"violations-index\" style=\"background-color: red; color: black; position: absolute;\">" + index + "</span>'", element);
    }

    private String getImpactColor(Impact impact) {
        switch (impact) {
            case CRITICAL:
                return "#da1e28";
            case SERIOUS:
                return "#ff832b";
            case MODERATE:
                return "#f1c21b";
            case MINOR:
                return "#add8e6";
            default:
                throw new IllegalArgumentException("Unknown impact");
        }
    }

    private byte[] takeScreenshot(WebDriver driver) {
        if(driver instanceof ChromeDriver dr) {
            Map<String, Object> layoutMetrics = ((ChromeDriver) driver).executeCdpCommand("Page.getLayoutMetrics", Collections.emptyMap());
            Map<String, Object> screenshotConfig = Maps.newHashMap();
            screenshotConfig.put("captureBeyondViewport", true);
            screenshotConfig.put("fromSurface", true);
            Map contentSize = (Map)layoutMetrics.get("cssContentSize");
            screenshotConfig.put("clip", ImmutableMap.of(
                    "width", contentSize.get("width"),
                    "height", contentSize.get("height"),
                    "x", 0,
                    "y", 0,
                    "scale", 0.5
            ));
            Map<String, Object> res = ((ChromeDriver) driver).executeCdpCommand("Page.captureScreenshot", screenshotConfig);
            return Base64.getDecoder().decode(res.get("data").toString());
        }
        throw new RuntimeException("Only supprted for ChromeDriver");
    }
}
